src/lua/core/Import.cpp

```diff
@@ -48,17 +95,267 @@ static std::string get_caller(lua_State *L)
+* a non-separating period.
+* This function discards the trailing component of the directory path. If you
+* wish to retain it, ensure the path string ends with '/'.
+*/
+static std::string path_to_module(std::string path)
+{
+	std::string module_name;
+	// simple allocation optimizations
+	if (path.size() > 8 && path.find_first_of('/') != std::string::npos) {
+		module_name.reserve(8);
+	}
+
+	// Loop through each path fragment and append to the module name
+	// e.g. '/data/modules/path/to/file.lua' would become 'data.modules.path.to'
+	size_t start = 0;
+	size_t end = path.find_first_of('/', 1); // use start=1 to generate string::npos when the path == "/"
+
+	while (end != std::string::npos) {
+		auto fragment = path.substr(start, end - start);
+
+		if (fragment.size() > 0) {
+			// Paths containing the separator character cannot be translated into a module name
+			// this means you can't do implicit lookup using a path starting in a hidden directory, e.g.
+			// /data/.local/mylib.lua
+			if (fragment.find('.') != std::string::npos)
+				return "";
+			module_name += module_name.empty() ? fragment : '.' + fragment;
+		}
+
+		start = end + 1;
+		end = path.find_first_of('/', start);
+	}
+
+	return module_name;
+}
+
+// If the given file path is cached, returns the cached object on the stack.
+// Otherwise if the path points to a valid file, executes that file and sets
+// the result in the module cache.
+// In all cases, if this function returns false the stack is unmodified.
+static bool load_file_cached(lua_State *L, const std::string &path, int cacheIdx)
+{
+	LUA_DEBUG_START(L);
+	cacheIdx = lua_absindex(L, cacheIdx);
+
+	// If the object is in the import cache, return it.
+	lua_getfield(L, cacheIdx, path.c_str());
+	if (!lua_isnil(L, -1)) {
+		DEBUG_INDENTED_PRINTF("-> loaded module file %s from cache.\n", path.c_str());
+		LUA_DEBUG_END(L, 1);
+		return true;
+	} else
+		lua_pop(L, 1);
+
+	// FileSystem::JoinPathBelow can't join a path starting with '/' for some reason
+	// TODO: find out why and remove that restriction.
+	std::string filePath = (path[0] == '/') ? path.substr(1) : path;
+	auto fileInfo = FileSystem::gameDataFiles.Lookup(filePath);
+	if (fileInfo.Exists()) {
+		// Load and run the file from disk
+		auto data = fileInfo.Read();
+		pi_lua_dofile(L, *data, 1);
+
+		DEBUG_INDENTED_PRINTF("-> loaded module file %s from disk\n", path.c_str());
+		DEBUG_INDENTED_PRINTF("-> module returned a %s value\n", lua_typename(L, lua_type(L, -1)));
+
+		// Since we loaded the module, we don't want to re-load it again.
+		// Replace the nil return value with 'true' to signify a successful
+		// module load
+		if (lua_isnil(L, -1)) {
+			lua_pop(L, 1);
+			lua_pushboolean(L, true);
+		}
+
+		// Store the module in the import cache
+		lua_pushvalue(L, -1);
+		lua_setfield(L, cacheIdx, path.c_str());
+
+		LUA_DEBUG_END(L, 1);
+		return true;
+	}
+
+	LUA_DEBUG_END(L, 0);
+	return false;
```

Comment:
Minor nitpick: If you're to have multiple returns in the function, I find it more readable to have the common case in the main indentation level and the error cases in the indented blocks (early return pattern). You'd have

```cpp
if (!fileInfo.Exists()) {
    LUA_DEBUG_END(L, 0);
    return false;
}

auto data = fileInfo.Read();
//etc...
```